### PR TITLE
Add Gradle generator

### PR DESCRIPTION
**Pull request by hadess**
I don't know what I'm doing, but I can parse the logs of a Gradle build and generate a `json` file to build Ghidra.

### CODE REVIEW

File: gradle/flatpak-gradle-generator.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                    break
+                sha256.update(data)
+    logging.info(f"done sha256({url})")
+    return sha256.hexdigest()
+
+async def parse_url(url):
+    ret = [{ 'type': 'file',
+            'url': url,
+            'sha256': await get_remote_sha256(url),
+            'dest': 'dependencies/flatRepo', }]
```
Comment (mpi3d) [inline gradle/flatpak-gradle-generator.py:29]: `dependencies/flatRepo` directory is specific to Ghidra only:
+ https://github.com/NationalSecurityAgency/ghidra/blob/357ff97cb3e845caa91499c3c334629ee5b34c00/build.gradle#L56
+ https://github.com/NationalSecurityAgency/ghidra/blob/357ff97cb3e845caa91499c3c334629ee5b34c00/build.gradle#L70-L80

So we need to add a "dependencies path" in the script args.

File: gradle/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+```sh
+$ source /usr/lib/sdk/openjdk11/enable.sh
+$ rm -rf gradle-cache
+$ mkdir -p dependencies/flatRepo/
+
+# Install some ghidra specific files, should probably be installed by hand/outside this script
+$ wget https://github.com/pxb1988/dex2jar/releases/download/2.0/dex-tools-2.0.zip
+$ unzip -j dex-tools-2.0.zip "*.jar" -d dependencies/flatRepo/
+$ wget -P dependencies/flatRepo/ https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/android4me/AXMLPrinter2.jar
```
Comment (mpi3d) [inline gradle/README.md:30]: This is also specific to Ghidra:
+ https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#manual-download-instructions

**Review by mpi3d — COMMENTED** [review]
Sounds good. I'll try to improve this.

File: gradle/flatpak-gradle-generator.py (hunk 1 above)
Comment (hadess) [reply gradle/flatpak-gradle-generator.py]: Done.

File: gradle/README.md (hunk 2 above)
Comment (hadess) [reply gradle/README.md:30]: This is explained before this block:
> then do what's needed to compile the application, **for example, for Ghidra**

**Comment by mpi3d** [comment]
The script doesn't work out of the box. I had to make some changes:
[flatpak-gradle-generator.zip](https://github.com/flatpak/flatpak-builder-tools/files/8671909/flatpak-gradle-generator.zip)
Also it supports `.exe` deps

**Comment by mpi3d** [comment]
Here it is:
[files.zip](https://github.com/flatpak/flatpak-builder-tools/files/8675636/files.zip)

**Comment by hadess** [comment]
I kept the code to create arch-specific urls, but I also:
- changed the arguments parsing to match the help (it wasn't possible to request multiple arches without changing the order or arguments in a way that didn't match the help output)
- used flatpak arches instead of the "gradle" ones in the arguments
- modified regex to not catch local filenames (instead of expecting "Downloading" etc.)
- add only-arches info to urls that need it

I'm pretty happy with this now, so if anyone else wants to experiment.

**Comment by hadess** [comment]
> Good. Any idea when this will be merged?

No, but porting the jadx or Bisq Flatpaks to use it would certainly show whether it's fit for purpose. In the meanwhile, it's not disappearing, and we can use it for Ghidra.

**Comment by Mailaender** [comment]
I couldn't get the script to work, but I don't understand the approach. For @terasology I can `./gradlew dependencies` which gives me:

```
Root project 'Terasology'
------------------------------------------------------------

api (n)
No dependencies

apiElements (n)
No dependencies

archives - Configuration for archive artifacts. (n)
No dependencies

classpath
No dependencies

codeMetrics
\--- org.terasology.config:codemetrics:1.6.3

default - Configuration for default artifacts. (n)
No dependencies

enforcedApiElements (n)
No dependencies

enforcedRuntimeElements (n)
No dependencies

natives
+--- org.lwjgl:lwjgl-bom:3.2.3
|    +--- org.lwjgl:lwjgl:3.2.3 (c)
|    +--- org.lwjgl:lwjgl-assimp:3.2.3 (c)
|    +--- org.lwjgl:lwjgl-glfw:3.2.3 (c)
|    +--- org.lwjgl:lwjgl-openal:3.2.3 (c)
|    +--- org.lwjgl:lwjgl-opengl:3.2.3 (c)
|    \--- org.lwjgl:lwjgl-stb:3.2.3 (c)
+--- org.lwjgl:lwjgl -> 3.2.3
+--- org.lwjgl:lwjgl-assimp -> 3.2.3
|    \--- org.lwjgl:lwjgl:3.2.3
+--- org.lwjgl:lwjgl-glfw -> 3.2.3
|    \--- org.lwjgl:lwjgl:3.2.3
+--- org.lwjgl:lwjgl-openal -> 3.2.3
|    \--- org.lwjgl:lwjgl:3.2.3
+--- org.lwjgl:lwjgl-opengl -> 3.2.3
|    \--- org.lwjgl:lwjgl:3.2.3
+--- org.lwjgl:lwjgl-stb -> 3.2.3
|    \--- org.lwjgl:lwjgl:3.2.3
+--- org.terasology.jnlua:jnlua_natives:0.1.0-SNAPSHOT
\--- org.terasology.jnbullet:JNBullet:1.0.2
```

**Comment by mpi3d** [comment]
I tried on Ghidra. But it doesn't list all the dependencies.

**Comment by hadess** [comment]
> I couldn't get the script to work, but I don't understand the approach.

What failed exactly? Can you provide the gradle log?

**Comment by Mailaender** [comment]
I am missing Python dependencies on @openSUSE.

**Comment by mpi3d** [comment]
Which ones? `aiohttp`? Just do a `pip install aiohttp`

**Comment by Mailaender** [comment]
Seems like updating pip with pip itself solved the installation problems. Yet the script does not run:

```
Traceback (most recent call last):
  File "./flatpak-gradle-generator.py", line 109, in <module>
    main()
  File "./flatpak-gradle-generator.py", line 101, in main
    sources = asyncio.run(parse_urls(urls, urls_arch, args.destdir))
AttributeError: module 'asyncio' has no attribute 'run'
```

**Comment by hadess** [comment]
> AttributeError: module 'asyncio' has no attribute 'run'

From https://docs.python.org/3/library/asyncio-task.html#asyncio.run
>  asyncio.run(coro, *, debug=False)
> New in version 3.7.

Python 3.7 was released in 2018, close to 4 years ago. Time to update.

**Comment by TobTobXX** [comment]
Just for some external input: I've also worked on this and have a semi-working script here: https://github.com/TobTobXX/flatpak-builder-tools/blob/master/gradle/flatpak-gradle-generator.sh.

Parsing build logs or `./gradlew dependencies` both have the issue that they don't include dependencies that are pure POMs (ie. parent POMs that are used for inheritance and referenced external POMs).

Instead, my script runs a `./gradle build --dry-run`, but sets a custom cache directory. This ensures all assets that are needed for an offline build are downloaded. (One issue my script has to deal with now is that some POMs differ depending on the mirror you use.)

I've not submitted a PR yet, since I'm still testing my script on various projects, but if you want to use it, go ahead.

**Comment by hadess** [comment]
> Parsing build logs or `./gradlew dependencies` both have the issue that they don't include dependencies that are pure POMs (ie. parent POMs that are used for inheritance and referenced external POMs).

I think this might be a problem with using gradlew instead of gradle, or a problem with the build scripts because the build tools should already log all the downloaded assets.

But not knowing what software you're trying to package, or the exact steps, it's difficult to say. Note that the build script we use for Ghidra uses a sandboxed dependency download section, to avoid any sort of cache from being used inadvertently.

> I've not submitted a PR yet, since I'm still testing my script on various projects, but if you want to use it, go ahead.

I doubt that a bash script would be accepted. In any case, that would be something to discuss on your own PR, not this one.

**Comment by Mailaender** [comment]
Can you maybe post an example usage? I managed to download all dependencies for JSkat, but it seems they are not discovered by Gradle:

```
Plugin [id: 'org.openjfx.javafxplugin', version: '0.0.13'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Plugin Repositories (could not resolve plugin artifact 'org.openjfx.javafxplugin:org.openjfx.javafxplugin.gradle.plugin:0.0.13')
  Searched in the following repositories:
    Gradle Central Plugin Repository
```

**Comment by hadess** [comment]
There's already an explanation in the README added in the patch. But I tested the generator with https://github.com/b0n541/jskat-multimodule (next time, please make sure to attach the full output of what you did, here I had to even search what jskat was).
```
$ flatpak install -y --user org.freedesktop.Sdk.Extension.openjdk17//21.08
[...]
$ flatpak run --command=bash --share=network --filesystem=`pwd` -d org.freedesktop.Sdk//21.08
[📦 org.freedesktop.Sdk jskat-multimodule]$ source /usr/lib/sdk/openjdk17/enable.sh
[📦 org.freedesktop.Sdk jskat-multimodule]$ ./gradlew -g gradle-cache/ --info --console plain clean installDist > gradle-log.txt
```

Running the script on the log threw an error, but I edited the script to remove the [URL that caused the error](https://plugins.gradle.org/m2/org/openjfx/javafxplugin/org.openjfx.javafxplugin.gradle.plugin/0.0.13/org.openjfx.javafxplugin.gradle.plugin-0.0.13.jar), and re-ran it, and it generated a json file. It didn't seem like the file was needed, as it built without it.

**Comment by Mailaender** [comment]
Running the generator is not a problem, but I don't know how to apply this in the Flatpak manifest.

**Comment by hadess** [comment]
> Running the generator is not a problem, but I don't know how to apply this in the Flatpak manifest.

It's a json file, you include it. See the Ghidra manifest for example:
https://github.com/flathub/org.ghidra_sre.Ghidra/blob/master/org.ghidra_sre.Ghidra.json#L142

**Comment by Mailaender** [comment]
The shell script moves files by default into `maven-local` where Gradle can't find it.

**Comment by hadess** [comment]
> The shell script moves files by default into `maven-local` where Gradle can't find it.

I'm not sure where you get that, there's no mentions of `maven-local` in the script.

> Does it need to be merged in order to use it?

Absolutely not, a second user might make it easier to get this patch merged, hopefully.

**Comment by MessiasLima** [comment]
I tried to run in my application and I'm using [detekt](https://detekt.dev/) for static code analysis. However, when I run it cannot get the `.jar` file from detekt because there is no `.jar` file in the repo. As you can see [here](https://plugins.gradle.org/m2/io/gitlab/arturbosch/detekt/io.gitlab.arturbosch.detekt.gradle.plugin/1.22.0/).

The script fails because of that but detekt is not even used in production code.

Do you have any hint about what to do in this case?

**Comment by mpi3d** [comment]
You might implement an option in the script to use a custom url for specific packages. Or even ignore the url.

Or you can currently remove the url manually.

**Comment by MessiasLima** [comment]
> You might implement an option in the script to use a custom url for specific packages. Or even ignore the url.
> 
> Or you can currently remove the url manually.

I ignored the Url and it worked, thank you!

**Comment by TingPing** [comment]
Please have somebody in CODEOWNERS, otherwise :+1:

**Comment by hadess** [comment]
> Please have somebody in `CODEOWNERS`, otherwise 👍

Done.

**Pull request merged**